### PR TITLE
lib/storage: enhancements for snapshots process

### PR DESCRIPTION
**Pull request by zekker6**
- Added timeout for snapshot creation
- Remove incomplete snapshot contents from disk
- Add metrics to track snapshots status

See related https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3551

### CODE REVIEW

File: app/vmstorage/main.go [hunk 1]
```diff
@@ -401,7 +408,11 @@ var (
 	staleSnapshotsRemoverWG sync.WaitGroup
 )
 
-var activeForceMerges = metrics.NewCounter("vm_active_force_merges")
+var (
+	activeForceMerges     = metrics.NewCounter("vm_active_force_merges")
+	snapshotsFailedTotal  = metrics.NewCounter("vm_snapshots_failed_total")
+	snapshotsCreatedTotal = metrics.NewCounter("vm_snapshots_created_total")
```
Comment (hagen1778) [inline app/vmstorage/main.go:414]: wdyt about using `vm_http_requests_total` metric instead?
It used for listing other /admin actions like `delete/{}/prometheus/api/v1/admin/tsdb/delete_series` and also has `vmagent_http_request_errors_total` counterpart.
I was kinda expecting to find `/snapshot` label in these metrics but I didn't)
You can also cover /list and /delete path parts if you go this way.

File: app/vmstorage/main.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ var (
 	forceMergeAuthKey = flag.String("forceMergeAuthKey", "", "authKey, which must be passed in query string to /internal/force_merge pages")
 	forceFlushAuthKey = flag.String("forceFlushAuthKey", "", "authKey, which must be passed in query string to /internal/force_flush pages")
 	snapshotsMaxAge   = flagutil.NewDuration("snapshotsMaxAge", "0", "Automatically delete snapshots older than -snapshotsMaxAge if it is set to non-zero duration. Make sure that backup process has enough time to finish the backup before the corresponding snapshot is automatically deleted")
+	snapshotTimeout   = flagutil.NewDuration("snapshotTimeout", "0", "Defines timeout value for process of creating new snapshot if it is set to non-zero duration. If set, make sure that timeout is lower than backup period.")
```
Comment (hagen1778) [inline app/vmstorage/main.go:31]:
```suggestion
	snapshotCreateTimeout   = flagutil.NewDuration("snapshotCreateTimeout", "0", "Defines timeout value for process of creating new snapshot. If set, make sure that timeout is lower than backup period.")
```

File: lib/storage/storage.go [hunk 3]
```diff
@@ -324,41 +324,48 @@ func (s *Storage) CreateSnapshot() (string, error) {
 		return "", fmt.Errorf("cannot create dir %q: %w", dstDataDir, err)
 	}
 
-	smallDir, bigDir, err := s.tb.CreateSnapshot(snapshotName)
+	smallDir, bigDir, err := s.tb.CreateSnapshot(snapshotName, deadline)
 	if err != nil {
+		fs.MustRemoveAll(dstDir)
```
Comment (hagen1778) [inline lib/storage/storage.go:329]: nit: what about wrap the func body into another func and check for returned error only once?

File: lib/mergeset/table.go [hunk 4]
```diff
@@ -1543,7 +1546,14 @@ func (tb *Table) CreateSnapshotAt(dstDir string) error {
 	if err != nil {
 		return fmt.Errorf("cannot read directory: %w", err)
 	}
-	for _, fi := range fis {
+
+	for i, fi := range fis {
+		if deadline > 0 && i%5 == 0 {
```
Comment (hagen1778) [inline lib/mergeset/table.go:1551]: are you sure that time is spent particularly in this loop?

Comment (zekker6) [reply lib/mergeset/table.go:1551]: I've checked the code paths for snapshot creation and this is one of the places where we can do deadline check and which takes some time.
Another time-wise expensive operation will be `fsync` syscall, but it does not have an option to cancel it.

**Review by valyala — APPROVED** [review]

**Review by valyala — APPROVED** [review]

**Comment by valyala** [comment]
FYI, it has been appeared that the snapshot process could take long time because the loop inside the `flushInmemoryParts()` function could be potentially infinite under high data ingestion rate. This has been fixed in the commit 6460475e3bb488086046cb5fdce3915206d731ef .

**Pull request merged**